Guidelines and processes to run OGC API - Processes / CWL / EOAP (#520)
Open-EO/openeo-processes: meta/implementation.md

- Julia: [Statistics.quantile](https://docs.julialang.org/en/v1/stdlib/Statistics/#Statistics.quantile!), type 7 is the default.
- Python: [numpy](https://numpy.org/doc/stable/reference/generated/numpy.quantile.html), [pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.quantile.html), [xarray](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.quantile.html) - type 7 (called 'linear' for the interpolation parameter) is the default for all of them.
- R: [quantile](https://stat.ethz.ch/R-manual/R-patched/library/stats/html/quantile.html) - type 7 is the default.

## OGC API - Processes

> OGC API - Processes and OGC EO Application Packages (AP) can generally be unilized in openEO in three different ways:

Reviewer: unilized?

Author: ```suggestion
OGC API - Processes and OGC EO Application Packages (AP) can generally be utilized in openEO in three different ways:
```

1. **openEO process**
  
   As a pre-defined process that exposes itself as a normal openEO process.
   It is not exposed to the user that in the background an AP is invoked.
2. **Pre-deployment through ADES**
  

>    In addition to the openEO API, a provider can offer access to an Application Deployment and Execution Service (ADES).

Author: ```suggestion
   In addition to the openEO API, a provider can offer access to an [Application Deployment and Execution Service (ADES)](https://github.com/eoepca/proc-ades).
```

Reviewer 2: So, probably we should cite this instead since the project was moved https://github.com/ZOO-Project/ZOO-Project/blob/4e0230a50e1afb64501a0853a3e2b23c5b9add19/docs/kernel/dru.rst

Author: @clausmichele I've updated the PR accordingly, see https://github.com/Open-EO/openeo-processes/pull/520/commits/e7ba6866d4b3691342433ca4bf6c0189c2fb4266

   {
     "process_id": "run_ogcapi",
     "arguments": {
       "data": ..., // Data, e.g. subtypes datacube or stac
       "id": "my-ap", // Identifier of the application package in OGC API - Processes
       "context": { // Parameters as defined in the CWL file
         "cwl_param1": true,
         "param2": 99
       }
     }
   }
   ```
3. **CWL provided at runtime (UDF runtime)**
  
   Providers can also provide a UDF runtime for the language CWL (instead of e.g. Python or R).
   The runtime is exposed through the endpoint `GET /udf_runtimes`.
  
   Example process node:
  
   ```json5
   {
     "process_id": "run_udf",
     "arguments": {
       "data": ..., // Data, e.g. subtypes datacube or stac
       "udf": "...", // CWL as YAML string/JSON object, URL, or file on the API user workspace
       "runtime": "cwl", // Assuming the UDF runtime is named "cwl"
       "context": { // Parameters as defined in the CWL file
         "cwl_param1": true,
         "param2": 99
       }
     }
   }
   ```

Generally, we recommend to use the following types and formats for the CWL inputs and outputs:

- `type`: `File` or `File[]` depending on the capabilities of the CWL workflow

> - `format`: For STAC inputs and outputs either `stac-item`, `stac-collection`, `stac-catalog`, `stac-itemcollection`, or `stac-stac` (i.e. and of the types before).

Author: We need to update the formats here with what is defined here:
https://github.com/opengeospatial/ogcapi-processes/pull/467

> - `format`: For STAC inputs and outputs either `stac-item`, `stac-collection`, `stac-catalog`, `stac-itemcollection`, or `stac-stac` (i.e. and of the types before).

Reviewer 2: @m-mohr typo? _(i.e. **and** of the types before)_ probably it should be **any**

Author: ```suggestion
- `format`: For STAC inputs and outputs either `stac-item`, `stac-collection`, `stac-catalog`, `stac-itemcollection`, or `stac-stac` (i.e. any of the types before).
```

> - `format`: For STAC inputs and outputs either `stac-item`, `stac-collection`, `stac-catalog`, `stac-itemcollection`, or `stac-stac` (i.e. any of the types before).

Author: ```suggestion
- `format`: For STAC inputs and outputs either:
  - `stac`: any of the following STAC entities
  - `stac-catalog`: STAC Catalog
  - `stac-collection`: STAC Collection
  - `stac-item`: STAC Item
  - `stac-item-collection`: STAC (API) ItemCollection
```
